core/arch/riscv64/mangle.c

```diff
@@ -784,10 +817,89 @@ mangle_exclusive_store(dcontext_t *dcontext, instrlist_t *ilist, instr_t *instr,
+ * the corresponding cache line. But this also puts many restrictions for instructions
+ * between LR/SC. For example, memory access instructions are not allowed.
+ *
+ * (Read more on the Volume I: RISC-V Unprivileged ISA V20191213 at page 51.)
+ *
+ * This is essentially the same situation as ARM/AArch64's exclusive monitors, quote from
+ * ldstex.dox: "Since dynamic instrumentation routinely adds additional memory loads and
+ * stores in between application instructions, it is in danger of breaking every monitor
+ * in the application."
+ *
+ * On a Unmatched RISC-V SBC, without this mangling, any application linked with libc
+ * would hang on startup.
+ *
+ * So for the LR/SC sequence, a similar approach to AArch64's exclusive monitors is
+ * adopted: mangling LR to a normal load, and SC to a compare-and-swap.
+ *
+ * While this introduces ABA problems, quote again from ldstex.dox: "the difference almost
+ * never matters for real programs".
+ *
+ * Here is an example of how we do the transformation:
+ *
+ * # Original code sequence
+ * 1:
+ *      lr.w.aqrl   a5, (a3)
+ *      bne         a5, a4, 1f
+ *      sc.w.rl     a1, a2, (a3)
+ *      bnez        a1, 1b
+ * 1:
+ *
+ * # After mangling
+ * <block 1>
+ * 1:
+ *      sd          a0, a0_slot(t3)         # save scratch register
+ *      fence       rw, rw                  # keep release semantics
+ *      lw          a5, 0(a3)               # replace lr with a normal load
+ *      fence       rw, rw                  # keep acquire semantics
+ *      sd          a3, lrsc_addr_slot(t3)  # save address
+ *      sd          a5, lrsc_val_slot(t3)   # save value
+ *      li          a0, 4
+ *      sd          a0, lrsc_size_slot(t3)  # save size (4 bytes)
+ *      ld          a0, a0_slot(t3)         # restore scratch register
+ *      bne         a5, a4, 1f
+ *
+ * <block 2>
+ * 1:
+ *      sd          a0, a0_slot(t3)         # save scratch register 1
+ *      sd          a4, a4_slot(t3)         # save scratch register 2
+ *      ld          a0, lrsc_addr_slot(t3)  # load saved address
+ *      bne         a0, a3, fail            # check address
+ *      ld          a0, lrsc_size_slot(t3)  # load saved size
+ *      li          a4, 4
+ *      bne         a0, a4, fail            # check size
+ *      ld          a0, lrsc_val_slot(t3)   # load saved value
+ * loop:
+ *      lr.w.aqrl   a4, (a3)                # begin of the CAS sequence
+ *      bne         a0, a4, final
+ *      sc.w.rl     a1, a2, (a3)
+ *      bne         a1, zero, loop          # retry on failure, end of the sequence
+ *      j           final
+ * fail:
+ *      li          a1, 1                   # sets non-zero value to dst on failure
+ * final:
+ *      li          a0, -1
+ *      sd          a0, lrsc_addr_slot(t3)  # invalidate reservation
+ *      ld          a0, a0_slot(t3)         # restore scratch register 1
+ *      ld          a4, a4_slot(t3)         # restore scratch register 2
+ */
```

Comment:
(Nice, I think these comments really help.). (Trying to use parens to mark as "no action requested"...wish github reviews had that as a real option.)